File: samples/android/AdUnitsSample/src/main/java/com/facebook/samples/AdUnitsSample/adapters/NativeAdRecyclerAdapter.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.samples.AdUnitsSample.adapters;

import android.app.Activity;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TextView;
import androidx.recyclerview.widget.RecyclerView;
import com.facebook.ads.Ad;
import com.facebook.ads.AdError;
import com.facebook.ads.AdOptionsView;
import com.facebook.ads.MediaView;
import com.facebook.ads.NativeAd;
import com.facebook.ads.NativeAdLayout;
import com.facebook.ads.NativeAdListener;
import com.facebook.ads.NativeAdsManager;
import com.facebook.samples.AdUnitsSample.R;
import com.facebook.samples.AdUnitsSample.models.RecyclerPostItem;
import java.util.ArrayList;
import java.util.List;

public class NativeAdRecyclerAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
  private List<RecyclerPostItem> mPostItems;
  private List<NativeAd> mAdItems;
  private NativeAdsManager mNativeAdsManager;
  private Activity mActivity;

  private static final int AD_DISPLAY_FREQUENCY = 5;
  private static final int POST_TYPE = 0;
  private static final int AD_TYPE = 1;

  private static final String TAG = "NativeAdsManager";

  public NativeAdRecyclerAdapter(
      Activity activity, List<RecyclerPostItem> postItems, NativeAdsManager nativeAdsManager) {
    mNativeAdsManager = nativeAdsManager;
    mPostItems = postItems;
    mAdItems = new ArrayList<>();
    mActivity = activity;
  }

  @Override
  public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
    if (viewType == AD_TYPE) {
      NativeAdLayout inflatedView =
          (NativeAdLayout)
              LayoutInflater.from(parent.getContext())
                  .inflate(R.layout.native_ad_unit, parent, false);
      return new AdHolder(inflatedView);
    } else {
      View inflatedView =
          LayoutInflater.from(parent.getContext())
              .inflate(R.layout.recycler_post_item, parent, false);
      return new PostHolder(inflatedView);
    }
  }

  @Override
  public int getItemViewType(int position) {
    return position % AD_DISPLAY_FREQUENCY == 0 ? AD_TYPE : POST_TYPE;
  }

  @Override
  public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
    if (holder.getItemViewType() == AD_TYPE) {
      NativeAd ad;

      if (mAdItems.size() > position / AD_DISPLAY_FREQUENCY) {
        ad = mAdItems.get(position / AD_DISPLAY_FREQUENCY);
      } else {
        ad =
            mNativeAdsManager.nextNativeAd(
                new NativeAdListener() {
                  @Override
                  public void onMediaDownloaded(Ad ad) {
                    // ignore
                  }

                  @Override
                  public void onError(Ad ad, AdError error) {
                    // ignore
                  }

                  @Override
                  public void onAdLoaded(Ad ad) {
                    // ignore
                  }

                  @Override
                  public void onAdClicked(Ad ad) {
                    Log.i(TAG, ((NativeAd) ad).getAdvertiserName() + " Ad Clicked");
                  }

                  @Override
                  public void onLoggingImpression(Ad ad) {
                    Log.i(TAG, ((NativeAd) ad).getAdvertiserName() + " Ad Impression");
                  }
                });
        if (!ad.isAdInvalidated()) {
          mAdItems.add(ad);
        } else {
          Log.w(NativeAdRecyclerAdapter.class.getSimpleName(), "Ad is invalidated!");
        }
      }

      AdHolder adHolder = (AdHolder) holder;
      adHolder.adChoicesContainer.removeAllViews();

      if (ad != null) {

        adHolder.tvAdTitle.setText(ad.getAdvertiserName());
        adHolder.tvAdBody.setText(ad.getAdBodyText());
        adHolder.tvAdSocialContext.setText(ad.getAdSocialContext());
        adHolder.tvAdSponsoredLabel.setText(R.string.sponsored);
        adHolder.btnAdCallToAction.setText(ad.getAdCallToAction());
        adHolder.btnAdCallToAction.setVisibility(
            ad.hasCallToAction() ? View.VISIBLE : View.INVISIBLE);
        AdOptionsView adOptionsView = new AdOptionsView(mActivity, ad, adHolder.nativeAdLayout);
        adHolder.adChoicesContainer.addView(adOptionsView, 0);

        List<View> clickableViews = new ArrayList<>();
        clickableViews.add(adHolder.ivAdIcon);
        clickableViews.add(adHolder.mvAdMedia);
        clickableViews.add(adHolder.btnAdCallToAction);
        ad.registerViewForInteraction(
            adHolder.nativeAdLayout, adHolder.mvAdMedia, adHolder.ivAdIcon, clickableViews);
      }
    } else {
      PostHolder postHolder = (PostHolder) holder;

      // Calculate where the next postItem index is by subtracting ads we've shown.
      int index = position - (position / AD_DISPLAY_FREQUENCY) - 1;

      RecyclerPostItem postItem = mPostItems.get(index);
      postHolder.tvPostContent.setText(postItem.getPostContent());
    }
  }

  @Override
  public int getItemCount() {
    return mPostItems.size() + mAdItems.size();
  }

  private static class PostHolder extends RecyclerView.ViewHolder {

    TextView tvPostContent;

    PostHolder(View view) {
      super(view);
      tvPostContent = view.findViewById(R.id.tvPostContent);
    }
  }

  private static class AdHolder extends RecyclerView.ViewHolder {

    NativeAdLayout nativeAdLayout;
    MediaView mvAdMedia;
    MediaView ivAdIcon;
    TextView tvAdTitle;
    TextView tvAdBody;
    TextView tvAdSocialContext;
    TextView tvAdSponsoredLabel;
    Button btnAdCallToAction;
    LinearLayout adChoicesContainer;

    AdHolder(NativeAdLayout adLayout) {
      super(adLayout);

      nativeAdLayout = adLayout;
      mvAdMedia = adLayout.findViewById(R.id.native_ad_media);
      tvAdTitle = adLayout.findViewById(R.id.native_ad_title);
      tvAdBody = adLayout.findViewById(R.id.native_ad_body);
      tvAdSocialContext = adLayout.findViewById(R.id.native_ad_social_context);
      tvAdSponsoredLabel = adLayout.findViewById(R.id.native_ad_sponsored_label);
      btnAdCallToAction = adLayout.findViewById(R.id.native_ad_call_to_action);
      ivAdIcon = adLayout.findViewById(R.id.native_ad_icon);
      adChoicesContainer = adLayout.findViewById(R.id.ad_choices_container);
    }
  }
}

File: samples/android/KotlinAdUnitsSample/src/main/java/com/facebook/samples/adunitssamplekotlin/adapters/NativeAdRecyclerAdapter.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.samples.adunitssamplekotlin.adapters

import android.app.Activity
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.LinearLayout
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.facebook.ads.*
import com.facebook.samples.adunitssamplekotlin.R
import com.facebook.samples.adunitssamplekotlin.models.RecyclerPostItem
import java.util.ArrayList

class NativeAdRecyclerAdapter(
    private val activity: Activity,
    private val postItems: List<RecyclerPostItem>,
    private val nativeAdsManager: NativeAdsManager
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
  private val adItems: MutableList<NativeAd>

  init {
    adItems = ArrayList()
  }

  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
    return if (viewType == AD_TYPE) {
      val inflatedView =
          LayoutInflater.from(parent.context).inflate(R.layout.native_ad_unit, parent, false)
              as NativeAdLayout
      AdHolder(inflatedView)
    } else {
      val inflatedView =
          LayoutInflater.from(parent.context).inflate(R.layout.recycler_post_item, parent, false)
      PostHolder(inflatedView)
    }
  }

  override fun getItemViewType(position: Int): Int {
    return if (position % AD_DISPLAY_FREQUENCY == 0) AD_TYPE else POST_TYPE
  }

  override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
    if (holder.itemViewType == AD_TYPE) {
      val ad: NativeAd?

      if (adItems.size > position / AD_DISPLAY_FREQUENCY) {
        ad = adItems[position / AD_DISPLAY_FREQUENCY]
      } else {
        ad = nativeAdsManager.nextNativeAd()
        if (ad != null && !ad.isAdInvalidated) {
          adItems.add(ad)
        } else {
          Log.w(NativeAdRecyclerAdapter::class.java.simpleName, "Ad is invalidated!")
        }
      }

      val adHolder = holder as AdHolder
      adHolder.adChoicesContainer.removeAllViews()

      ad?.let { nonNullAd ->
        adHolder.tvAdTitle.text = nonNullAd.advertiserName
        adHolder.tvAdBody.text = nonNullAd.adBodyText
        adHolder.tvAdSocialContext.text = nonNullAd.adSocialContext
        adHolder.tvAdSponsoredLabel.setText(R.string.sponsored)
        adHolder.btnAdCallToAction.text = nonNullAd.adCallToAction
        adHolder.btnAdCallToAction.visibility =
            if (nonNullAd.hasCallToAction()) View.VISIBLE else View.INVISIBLE
        val adOptionsView = AdOptionsView(activity, nonNullAd, adHolder.nativeAdLayout)
        adHolder.adChoicesContainer.addView(adOptionsView, 0)

        val clickableViews = ArrayList<View>()
        clickableViews.add(adHolder.ivAdIcon)
        clickableViews.add(adHolder.mvAdMedia)
        clickableViews.add(adHolder.btnAdCallToAction)
        nonNullAd.registerViewForInteraction(
            adHolder.nativeAdLayout, adHolder.mvAdMedia, adHolder.ivAdIcon, clickableViews)
      }
    } else {
      val postHolder = holder as PostHolder

      // Calculate where the next postItem index is by subtracting ads we've shown.
      val index = position - position / AD_DISPLAY_FREQUENCY - 1

      val postItem = postItems[index]
      postHolder.tvPostContent.text = postItem.postContent
    }
  }

  override fun getItemCount(): Int {
    return postItems.size + adItems.size
  }

  private class PostHolder internal constructor(view: View) : RecyclerView.ViewHolder(view) {
    internal var tvPostContent: TextView = view.findViewById(R.id.tvPostContent)
  }

  private class AdHolder internal constructor(internal var nativeAdLayout: NativeAdLayout) :
      RecyclerView.ViewHolder(nativeAdLayout) {
    internal var mvAdMedia: MediaView = nativeAdLayout.findViewById(R.id.native_ad_media)
    internal var ivAdIcon: MediaView = nativeAdLayout.findViewById(R.id.native_ad_icon)
    internal var tvAdTitle: TextView = nativeAdLayout.findViewById(R.id.native_ad_title)
    internal var tvAdBody: TextView = nativeAdLayout.findViewById(R.id.native_ad_body)
    internal var tvAdSocialContext: TextView =
        nativeAdLayout.findViewById(R.id.native_ad_social_context)
    internal var tvAdSponsoredLabel: TextView =
        nativeAdLayout.findViewById(R.id.native_ad_sponsored_label)
    internal var btnAdCallToAction: Button =
        nativeAdLayout.findViewById(R.id.native_ad_call_to_action)
    internal var adChoicesContainer: LinearLayout =
        nativeAdLayout.findViewById(R.id.ad_choices_container)
  }

  companion object {

    private const val AD_DISPLAY_FREQUENCY = 5
    private const val POST_TYPE = 0
    private const val AD_TYPE = 1
  }
}
